Parse and verify certificates with options

yogisinha opened the pull request:
<!--

Thanks for submitting a PR to OPA!

Before pressing 'Create pull request' please read the checklist below.

* All code changes should be accompanied with tests. If you are not
modifying any tests, just provide a short explanation of why updates
to tests are not necessary. In addition to helping catch bugs, tests
are extremely helpful in providing _context_ that explains how your
changes can be used.

* All changes to public APIs **must** be accompanied with
docs. Examples of public APIs include built-in functions,
config fields, and of course, exported Go types/functions/constants/etc.

* Commit messages should explain _why_ you made the changes, not what
you changed. Use active voice. Keep the subject line under 50
characters or so.

* All commits must be signed off by the author. If you are not
familiar with signing off, see our contributor guide below.

For more information on contributing to OPA see:

* [Contributing Guide](https://www.openpolicyagent.org/docs/latest/contributing/)
  for high-level contributing guidelines and development setup.

-->

### Why the changes in this PR are needed?
<!--
Include a short description of WHY the changes were made.
-->
parse and verify x509 certificates with additional options was requested

### What are the changes in this PR?
buitins.go and crypto.go were modified
<!--
Include a short description of WHAT changes were made.
-->

### Notes to assist PR review:
This is a draft PR as requested by @ashutosh-narkar.
Please comment if the implementation looks right 
<!--
Here you can add information you think will help the reviewer(s).
-->

### Further comments:
Just to pass the existing tests I created a another method `verifyX509CertificateChainWithOpts` which is same as `verifyX509CertificateChain` except it takes additional `verifyOptions` struct. I can remove this method and add this extra `verifyOptions` to `verifyX509CertificateChain` and change the test cases later. `verifyX509CertificateChainWithOpts` is just temporary.

For an e.g. This method will be called as :

```
crypto.x509.parse_and_verify_certificates_with_options(
pem encoded cert data,
'{
"DNSName": "some_value", "CurrentTime": "value in millis epoch", 
"KeyUsages": ["KeyUsageServerAuth", "KeyUsageClientAuth", "KeyUsageCodeSigning"]}",
"MaxConstraintComparisons": "int value"
)'
)
```
<!--
Here you can include links to additional resources related to the changes, discuss your solution, other approaches you considered etc.
-->

ashutosh-narkar commented:
@yogisinha thanks for the contribution. Can you please resolve the conflicts in the PR?

yogisinha commented:
I resolved the conflicts.

charlieegan3 commented on ast/builtins.go:
```diff
@@ -2327,6 +2328,37 @@ with all others being treated as intermediates.`,
+	Name: "crypto.x509.parse_and_verify_certificates_with_options",
+	Description: `Returns one or more certificates from the given string containing PEM
+or base64 encoded DER certificates after verifying the supplied certificates form a complete
+certificate chain back to a trusted root. Second argument is a option object to validate the 
```
This line looks to be incomplete? the note about the options seems to be cut off.

yogisinha replied:
Hi,
its giving me following error on last step. I didn't have `hugo` so I installed it by `sudo apt  install hugo`

~/opa/docs/website$ **hugo server --contentDir=generated/**
```
Building sites … ERROR 2024/02/21 11:55:13 render of "page" failed: "/home/yogi/opa/docs/website/layouts/docs/ecosystem-single.html.html:20:17": execute of template failed: template: docs/ecosystem-single.html.html:20:17: executing "content" at <path>: can't evaluate field BaseName in type interface {}
ERROR 2024/02/21 11:55:13 render of "page" failed: "/home/yogi/opa/docs/website/layouts/docs/ecosystem-single.html.html:20:17": execute of template failed: template: docs/ecosystem-single.html.html:20:17: executing "content" at <path>: can't evaluate field BaseName in type interface {}
ERROR 2024/02/21 11:55:13 render of "page" failed: "/home/yogi/opa/docs/website/layouts/docs/ecosystem-single.html.html:20:17": execute of template failed: template: docs/ecosystem-single.html.html:20:17: executing "content" at <path>: can't evaluate field BaseName in type interface {}
ERROR 2024/02/21 11:55:13 render of "page" failed: "/home/yogi/opa/docs/website/layouts/docs/ecosystem-single.html.html:20:17": execute of template failed: template: docs/ecosystem-single.html.html:20:17: executing "content" at <path>: can't evaluate field BaseName in type interface {}
Built in 1659 ms
Error: Error building site: failed to render pages: render of "page" failed: "/home/yogi/opa/docs/website/layouts/docs/ecosystem-single.html.html:20:17": execute of template failed: template: docs/ecosystem-single.html.html:20:17: executing "content" at <path>: can't evaluate field BaseName in type interface {}
```

charlieegan3 replied:
Hmm, I am not sure what would be causing that... 🤔 

how about here: https://deploy-preview-6583--openpolicyagent.netlify.app/docs/edge/policy-reference/#builtin-crypto-cryptox509parse_and_verify_certificates_with_options

charlieegan3 commented on ast/builtins.go:
```diff
@@ -2327,6 +2328,37 @@ with all others being treated as intermediates.`,
+						types.NewSet(types.A),
+					),
+				)),
+			).Description("object containing a set or array of neighboring vertices"),
```
> neighboring vertices

What is this meant to refer to. I thought the options were meant to be the key usages?

charlieegan3 commented on topdown/crypto.go:
```diff
@@ -122,6 +123,137 @@ func builtinCryptoX509ParseAndVerifyCertificates(_ BuiltinContext, operands []*a
 	return iter(valid)
 }
 
+type verifyOptions struct {
```
question: is there a reason we can't use the golang type? https://pkg.go.dev/crypto/x509#VerifyOptions

charlieegan3 commented on topdown/crypto.go:
```diff
@@ -122,6 +123,137 @@ func builtinCryptoX509ParseAndVerifyCertificates(_ BuiltinContext, operands []*a
+	return iter(valid)
+}
+
+func extractVerifyOpts(options ast.Object) (verifyOpt verifyOptions, err error) {
```
I think that it would be nice to have some tests for this function that map the Rego options to the golang struct (https://pkg.go.dev/crypto/x509#VerifyOptions)

charlieegan3 commented on topdown/crypto.go:
```diff
@@ -425,6 +558,41 @@ func verifyX509CertificateChain(certs []*x509.Certificate) ([]*x509.Certificate,
 	return chains[0], nil
 }
 
+func verifyX509CertificateChainWithOpts(certs []*x509.Certificate, vo verifyOptions) ([]*x509.Certificate, error) {
```
I think that rather than making a new function here we can extend verifyX509CertificateChain to support the passing of an existing options struct that is extended if present or initialised fresh if not.

charlieegan3 replied:
👍 sounds like a good plan, thanks.

charlieegan3 reviewed (commented):
Hi @yogisinha, thanks for this PR. It's a great start and having these options available will be handy.

In addition to my comments below, I think it'd be good to have some test cases for the new function. You can see some examples here to get you started https://github.com/open-policy-agent/opa/blob/cfb03ca495111cbcce98acce93bf96b8433f4f35/topdown/crypto_test.go#L144-L143

Thanks again and sorry to take a while to get back to you. that's on me.

yogisinha commented:
@charlieegan3 ,
I made all the changes. Please take a look.

charlieegan3 commented:
Hey, I have added a comment here https://github.com/open-policy-agent/opa/pull/6583#discussion_r1497884139 regarding previewing the docs. I think you're also going to need to run make generate and commit the new builtins JSON too. Thanks!

yogisinha commented:
> Hey, I have added a comment here [#6583 (comment)](https://github.com/open-policy-agent/opa/pull/6583#discussion_r1497884139) regarding previewing the docs. I think you're also going to need to run make generate and commit the new builtins JSON too. Thanks!

Added the json files.

charlieegan3 commented on ast/builtins.go:
```diff
@@ -2327,6 +2328,51 @@ with all others being treated as intermediates.`,
+	Description: `Returns one or more certificates from the given string containing PEM
+or base64 encoded DER certificates after verifying the supplied certificates form a complete
+certificate chain back to a trusted root. Second argument is a option object to pass
+extra configs to verify the validity of certificates.
```
"Second argument is a option object to pass extra configs to verify the validity of certificates."

I would change this to:

"A config option passed as the second argument can be used to configure the validation options used."

charlieegan3 commented on ast/builtins.go:
```diff
@@ -2327,6 +2328,51 @@ with all others being treated as intermediates.`,
+extra configs to verify the validity of certificates.
+
+The first certificate is treated as the root and the last is treated as the leaf,
+with all others being treated as intermediates.` +
```
```suggestion
with all others being treated as intermediates.

` +
```
<img width="232" alt="Screenshot 2024-02-21 at 18 11 56" src="https://github.com/open-policy-agent/opa/assets/1774239/829ce6e9-5af9-412a-8353-d7a5500322e4">

charlieegan3 commented on ast/builtins.go:
```diff
@@ -2327,6 +2328,51 @@ with all others being treated as intermediates.`,
+		"`\"MaxConstraintComparisons\" : 5,`" +
+		"`}`" +
+		"Then this function call will look like: `crypto.x509.parse_and_verify_certificates_with_options(<cert base64 encoded string>, options)`" +
+		"This option's field has same definitions as fields in [x509.VerifyOptions struct](https://pkg.go.dev/crypto/x509#VerifyOptions)" +
```
https://github.com/open-policy-agent/opa/assets/1774239/9a6e7ec8-f2c7-413f-b1fd-55545e6649bc

I think that we need to make some of these lines less long, as it's pushing out the rest of the table at the moment. Some strategic line breaks should be 👌

yogisinha replied:
Yes, I realized that too and was already working on that

charlieegan3 commented on ast/builtins.go:
```diff
@@ -2327,6 +2328,51 @@ with all others being treated as intermediates.`,
+		"`\"MaxConstraintComparisons\" : 5,`" +
+		"`}`" +
+		"Then this function call will look like: `crypto.x509.parse_and_verify_certificates_with_options(<cert base64 encoded string>, options)`" +
+		"This option's field has same definitions as fields in [x509.VerifyOptions struct](https://pkg.go.dev/crypto/x509#VerifyOptions)" +
```
```suggestion
		"This options object has the same fields as [x509.VerifyOptions](https://pkg.go.dev/crypto/x509#VerifyOptions)" +
```

charlieegan3 commented on ast/builtins.go:
```diff
@@ -2327,6 +2328,51 @@ with all others being treated as intermediates.`,
+	Decl: types.NewFunction(
+		types.Args(
+			types.Named("certs", types.S).Description("base64 encoded DER or PEM data containing two or more certificates where the first is a root CA, the last is a leaf certificate, and all others are intermediate CAs"),
+			types.Named("options", types.NewObject(
```
```suggestion
			types.Named("options", types.NewObject(nil, types.NewDynamicProperty(types.S, types.A)))
```

I think we can just make it a string to any object since there are various types in the struct.

charlieegan3 commented on ast/builtins.go:
```diff
@@ -2327,6 +2328,51 @@ with all others being treated as intermediates.`,
+		"`}`" +
+		"Then this function call will look like: `crypto.x509.parse_and_verify_certificates_with_options(<cert base64 encoded string>, options)`" +
+		"This option's field has same definitions as fields in [x509.VerifyOptions struct](https://pkg.go.dev/crypto/x509#VerifyOptions)" +
+		"`CurrentTime` is nanoseconds since epoch" +
```
```suggestion
		"`CurrentTime` is the number of nanoseconds since the Unix Epoch as a number" +
```

charlieegan3 commented on topdown/crypto.go:
```diff
@@ -122,6 +123,130 @@ func builtinCryptoX509ParseAndVerifyCertificates(_ BuiltinContext, operands []*a
+
+	for _, key := range options.Keys() {
+		switch key.String() {
+		case "\"DNSName\"":
```
It might be nicer to clean the quote from the key at the top of the loop iteration, this might make it look a little more tidy in the cases.

charlieegan3 commented on topdown/crypto.go:
```diff
@@ -122,6 +123,130 @@ func builtinCryptoX509ParseAndVerifyCertificates(_ BuiltinContext, operands []*a
+						}
+					}
+				})
+
```
```suggestion
```

charlieegan3 commented on topdown/crypto.go:
```diff
@@ -122,6 +123,130 @@ func builtinCryptoX509ParseAndVerifyCertificates(_ BuiltinContext, operands []*a
+				})
+
+			} else {
+				return verifyOpt, fmt.Errorf("'KeyUsages' should be a set")
```
Might we want to support the case where the user supplies an array?

charlieegan3 replied:
I was just imagining that a user might provide an array by mistake, if it were coming from input or data loaded from JSON this is quite likely. I agree that a Set is what we'd like to receive in an ideal case, but I think there's merit in covering an array as input too.

yogisinha replied:
makes sense. I would change that.

yogisinha replied:
I think having dup in array is ok. its not a issue. I looked into Go code and passing list with dup key usages should be fine. just extra loop executions in this part of code: https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/crypto/x509/verify.go;l=1166

yogisinha replied:
I added the Array type for KeyIsages.

charlieegan3 commented on topdown/crypto_test.go:
```diff
@@ -647,3 +654,91 @@ xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+	}{
+		{
+			jsonOption: ast.MustParseTerm(`{"DNSName": 1}`),
+			expectErr:  true,
```
It would be nice to know what the error was in such cases, perhaps checking a string or something in the err.Error()

charlieegan3 reviewed (commented):
Thanks for adding the test and the other edits based on my comments above. It's taking shape, just a few more minor ones.

charlieegan3 commented:
Hey @yogisinha, are you able to take a look at the conflicts with main again?

yogisinha commented:
> Hey @yogisinha, are you able to take a look at the conflicts with main again?

yogisinha commented:
Are you talking about this message "This branch is out-of-date with the base branch" . This was not coming before. What should I do ? Should I do "Update with merge commit" ?

charlieegan3 commented:
Hey, yeah best would be to run:

```
git fetch origin
git rebase origin/main
# address any conflicts and make sure the tests all pass etc
git rebase --continue
git push -f
```

yogisinha commented:
I don't see any merge conflicts . There is nothing to push (no files changed) when I did the above .

ashutosh-narkar commented:
@yogisinha can you please rebase this branch? I'm unable to do it via the UI. Similar to what Charlie mentioned [here](https://github.com/open-policy-agent/opa/pull/6583#issuecomment-1967754209).  Also make sure your `origin` points to `git@github.com:open-policy-agent/opa.git`

ashutosh-narkar commented on ast/builtins.go:
```diff
@@ -2327,6 +2328,44 @@ with all others being treated as intermediates.`,
+with all others being treated as intermediates.
+
+` +
+		"Example option object: " +
+		"`options := {\"DNSName\" : \"example.dns.com\", \"CurrentTime\": 1708447636000000000,`" +
+		"`\"KeyUsages\": {\"KeyUsageServerAuth\", \"KeyUsageClientAuth\", \"KeyUsageCodeSigning\"}}` " +
+		"Then this function call will look like: `crypto.x509.parse_and_verify_certificates_with_options(<cert base64 encoded string>, options)`. " +
+		"This options object has the same fields as [x509.VerifyOptions struct](https://pkg.go.dev/crypto/x509#VerifyOptions). " +
+		"`CurrentTime` is the number of nanoseconds since the Unix Epoch as a number. " +
+		"List of possible values for `KeyUsages` field are: `[\"KeyUsageAny\", \"KeyUsageServerAuth\",` " +
+		"`\"KeyUsageClientAuth\", \"KeyUsageCodeSigning\", \"KeyUsageEmailProtection\", \"KeyUsageIPSECEndSystem\",` " +
+		"`\"KeyUsageIPSECTunnel\", \"KeyUsageIPSECUser\", \"KeyUsageTimeStamping\", \"KeyUsageOCSPSigning\",` " +
+		"`\"KeyUsageMicrosoftServerGatedCrypto\", \"KeyUsageNetscapeServerGatedCrypto\",` " +
+		"`\"KeyUsageMicrosoftCommercialCodeSigning\", \"KeyUsageMicrosoftKernelCodeSigning\"]`",
```
I think it would be better to skip this example object and simply add a note to refer to the [x509.VerifyOptions struct](https://pkg.go.dev/crypto/x509#VerifyOptions). If you look at the deploy preview it looks too verbose. If folks have questions on usage we can always add in more details. WDYT?

ashutosh-narkar commented on ast/builtins.go:
```diff
@@ -2327,6 +2328,44 @@ with all others being treated as intermediates.`,
+			types.Named("options", types.NewObject(
+				nil,
+				types.NewDynamicProperty(types.S, types.A),
+			)).Description("option object to pass extra configs to verify the validity of certificates."),
```
Nit: What about `object containing extra configs to verify the validity of certificates`

ashutosh-narkar commented on topdown/crypto.go:
```diff
@@ -122,6 +123,138 @@ func builtinCryptoX509ParseAndVerifyCertificates(_ BuiltinContext, operands []*a
+
+func extractVerifyOpts(options ast.Object) (verifyOpt x509.VerifyOptions, err error) {
+
+	for _, key := range options.Keys() {
+		k := key.String()[1 : len(key.String())-1]
```
```suggestion
	for _, val := range options.Keys() {
		key, err := ast.JSON(val.Value)
		if err != nil {
			return verifyOpt, err
		}

		key = key.(string)
```

Then switch on `key`

yogisinha replied:
ok. great. will change that. did not know abt ast.JSON function.

ashutosh-narkar commented on topdown/crypto.go:
```diff
@@ -122,6 +123,138 @@ func builtinCryptoX509ParseAndVerifyCertificates(_ BuiltinContext, operands []*a
+		case "DNSName":
+			dns, ok := options.Get(key).Value.(ast.String)
+			if ok {
+				verifyOpt.DNSName = dns.String()[1 : len(dns.String())-1]
```
What about something like 👇 

```suggestion
				verifyOpt.DNSName = strings.Trim(string(dns), "\"")
```

ashutosh-narkar commented on topdown/crypto.go:
```diff
@@ -122,6 +123,138 @@ func builtinCryptoX509ParseAndVerifyCertificates(_ BuiltinContext, operands []*a
+			ks.Foreach(func(t *ast.Term) {
+				u, ok := t.Value.(ast.String)
+				if ok {
+					v := u.String()[1 : len(u.String())-1]
```
Same comment as above.

ashutosh-narkar commented on topdown/crypto.go:
```diff
@@ -122,6 +123,138 @@ func builtinCryptoX509ParseAndVerifyCertificates(_ BuiltinContext, operands []*a
+				u, ok := t.Value.(ast.String)
+				if ok {
+					v := u.String()[1 : len(u.String())-1]
+					if k, ok := allowedKeyUsages[v]; ok {
```
Do we need to check for invalid options here?

ashutosh-narkar reviewed (commented):
@yogisinha this is going is the right direction. Few comments inline. Thanks for working on this.

yogisinha commented:
@ashutosh-narkar , 
I made all the suggested changes.

ashutosh-narkar commented on ast/builtins.go:
```diff
@@ -2327,6 +2328,40 @@ with all others being treated as intermediates.`,
+with all others being treated as intermediates.
+
+` +
+		"`options` object has the same fields as [x509.VerifyOptions struct](https://pkg.go.dev/crypto/x509#VerifyOptions). " +
```
This gives the impression that we support all the `VerifyOptions` fields which is not the case, correct? 
We can just mention the values we support and point to the Go docs for more info. We can skip the below description as well.

yogisinha replied:
Made the change.

ashutosh-narkar commented on ast/builtins.go:
```diff
@@ -2327,6 +2328,41 @@ with all others being treated as intermediates.`,
 
+var CryptoX509ParseAndVerifyCertificatesWithOptions = &Builtin{
+	Name: "crypto.x509.parse_and_verify_certificates_with_options",
+	Description: `Returns one or more certificates from the given string containing PEM
```
Sorry about the back and forth on this. We can update the `Description` text part to:

```
Returns one or more certificates from the given string containing PEM or base64 encoded DER certificates after verifying the supplied certificates form a complete certificate chain back to a trusted root. A config option passed as the second argument can be used to configure the validation options used.

The first certificate is treated as the root and the last is treated as the leaf, with all others being treated as intermediates.

`options` object supports four fields which maps to same fields in [x509.VerifyOptions struct](https://pkg.go.dev/crypto/x509#VerifyOptions) namely `DNSName`, `CurrentTime`, `KeyUsages` and `MaxConstraintComparisions`.  
```

yogisinha replied:
In our options, 3 fields are kind of diff types than that in Go lib
`CurrentTime`: number
`KeyUsages`: list of strings,  Go lib ; list of int
`MaxConstraintComparisions`: the types are same but in Go lib its mis-spelled (.. `sions`) extra "i" before "o". Ours is `MaxConstraintComparisons`. Thats why I explained at least these 3 fields. if we don't explain them , user might get confused.

ashutosh-narkar replied:
Ok thanks for the clarification. We can keep it then.

ashutosh-narkar reviewed (commented):
@yogisinha thanks for working on this. It would be helpful to add a unit test that actually tests the new builtin itself in addition to what you already have. Thanks.

yogisinha commented:
Added the test case.

ashutosh-narkar reviewed (commented):
Thanks for adding the test cases and addressing other comments. Please squash your commits and we can get this in.

charlieegan3 commented on topdown/crypto.go:
```diff
@@ -122,6 +123,149 @@ func builtinCryptoX509ParseAndVerifyCertificates(_ BuiltinContext, operands []*a
+	input, err := builtins.StringOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
```
nit: add a new line after this if err check

charlieegan3 reviewed (commented):
Hi @yogisinha, thanks for all the work on this. I think the code is looking great now. 

My only final comment is that the docs are still pushing out the table. This built-in table is a bit of a pain to work with... but I'd like to keep it without a h-scroll bar if possible. Perhaps the options can be a bulleted list?

https://github.com/open-policy-agent/opa/assets/1774239/618e89d8-2ef2-4222-a739-c879159e99c0

yogisinha commented:
@charlieegan3 , 
I think it looks good now. I had to make several commits incrementally to see the behavior of deploy preview.

charlieegan3 commented on topdown/crypto.go:
```diff
@@ -122,6 +123,150 @@ func builtinCryptoX509ParseAndVerifyCertificates(_ BuiltinContext, operands []*a
+		if err != nil {
+			return verifyOpt, err
+		}
+		k = k.(string)
```
Since ast.JSON returns an `interface{}`, I think we need to make this a check too.

E.g. `key, ok := k.(string)`

and then `if !ok { ... // return err etc }`

charlieegan3 replied:
Ahh I see, hmm... perhaps we can just continue in the loop if the key is not a string?

yogisinha replied:
Added that.

charlieegan3 reviewed (commented):
This looks good to me now, I can still see that the table is overflowing but there are other overflowing tables on the page (`net` functions) so I think we need to do some more work on the site's styles there. 

I have one final comment, then we can rebase and get this in.

charlieegan3 commented:
Hey, I think we are nearly there now @yogisinha, I have updated the generated code in https://github.com/open-policy-agent/opa/pull/6583/commits/91e90078e79b579c938e3258cc2b0872a20f8bab but can you rebase this from origin/main so that we can get this in. I'm at KubeCon at the moment so have limited time to help, but will reply when I get a chance.

Thanks for your work here, much appreciated!

yogisinha commented:
Hi @charlieegan3 ,
I created a new PR . I was getting lot of conflicts when I was trying to rebase, don't know why. In that new PR, everything is in  single commit.

New PR: https://github.com/open-policy-agent/opa/pull/6643

ashutosh-narkar commented:
Closing this in favor of https://github.com/open-policy-agent/opa/pull/6643.
